### PR TITLE
#NEWRELIC-4037 Mongodb reactive streams driver

### PR DESCRIPTION
**Pull request by tbradellis**
Instruments mongdb reactive streams driver. (MongoCollection)
Still some work needed to finish and confirm good coverage across threads.

This was intended to target driver version 4.2.0+.  However, this approach may have some overlap with previous modules. We may need to either find a way to exclude instrumentation for those previous, or refactor the old so that this module that gets used for 4.0.0+.

### Overview
Describe the changes present in the pull request

### Related Github Issue
https://issues.newrelic.com/browse/NEWRELIC-4037
### Testing
Will need to build new instrumentation tests.

### CODE REVIEW

File: instrumentation/mongodb-reactive-streams-4.2/src/main/java/com/mongodb/reactivestreams/client/MongoCollection_Instrumentation.java [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+                    .collection(getNamespace().getCollectionName())
+                    .operation(operation)
+                    .noInstance()
+                    .databaseName(getNamespace().getDatabaseName())
+                    .build();
+            transaction.getTracedMethod().reportAsExternal(params);
+        }
+    }
+
+}
```
Comment (philleonard) [inline instrumentation/mongodb-reactive-streams-4.2/src/main/java/com/mongodb/reactivestreams/client/MongoCollection_Instrumentation.java:464]:
```suggestion
}

```

File: instrumentation/mongodb-reactive-streams-4.2/src/main/java/com/nr/agent/mongo/MongoUtil.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    public static final String OP_INSERT_ONE = "insertOne";
+    public static final String OP_INSERT_MANY = "insertMany";
+    public static final String OP_UPDATE_MANY = "updateMany";
+    public static final String OP_UPDATE_ONE = "updateOne";
+
+    public static final String OP_GET_MORE = "getMore";
+    public static final String OP_GROUP = "group";
+    public static final String UNKNOWN = "Unknown";
+
+    // "delete" commands are different from DBCollection.remove
```
Comment (philleonard) [inline instrumentation/mongodb-reactive-streams-4.2/src/main/java/com/nr/agent/mongo/MongoUtil.java:64]:
```suggestion
    /**
    * "delete" commands are different from DBCollection.remove.
    */
```
For consistency with the other comment lines below?

Perhaps also now we are using doc comments:
```suggestion
    /**
    * "delete" commands are different from {@link DBCollection#remove}.
    */
```

File: instrumentation/mongodb-reactive-streams-4.2/src/main/java/com/mongodb/reactivestreams/client/MongoCollection_Instrumentation.java [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+            final TDocument replacement,
+            final ReplaceOptions options) {
+
+        reportAsExternal(OP_REPLACE_ONE);
+        return Weaver.callOriginal();
+    }
+
+    @Trace(leaf = true)
+    public Publisher<UpdateResult> updateOne(final Bson filter, final Bson update, final UpdateOptions options) {
+
```
Comment (philleonard) [inline instrumentation/mongodb-reactive-streams-4.2/src/main/java/com/mongodb/reactivestreams/client/MongoCollection_Instrumentation.java:214]:
```suggestion
```
Same would apply elsewhere but I suspect it is a formatting rule for method signatures that are broken up onto multiple lines, of which this is not one of these. But not sure.

File: instrumentation/mongodb-reactive-streams-4.2/src/main/java/com/mongodb/reactivestreams/client/MongoCollection_Instrumentation.java [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+        return Weaver.callOriginal();
+
+    }
+
+    @Trace(leaf = true)
+    public Publisher<UpdateResult> updateOne(final Bson filter, final List<? extends Bson> update,
+            final UpdateOptions options) {
+        reportAsExternal(OP_UPDATE_ONE);
+        return Weaver.callOriginal();
+
```
Comment (philleonard) [inline instrumentation/mongodb-reactive-streams-4.2/src/main/java/com/mongodb/reactivestreams/client/MongoCollection_Instrumentation.java:232]:
```suggestion
```
Same applies below.

File: instrumentation/mongodb-reactive-streams-4.2/src/main/java/com/mongodb/reactivestreams/client/MongoCollection_Instrumentation.java [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+        reportAsExternal(OP_REPLACE_ONE);
+        return Weaver.callOriginal();
+
+    }
+
+    @Trace(leaf = true)
+    public Publisher<UpdateResult> replaceOne(final ClientSession clientSession, final Bson filter,
+            final TDocument replacement,
+            final ReplaceOptions options) {
+
```
Comment (philleonard) [inline instrumentation/mongodb-reactive-streams-4.2/src/main/java/com/mongodb/reactivestreams/client/MongoCollection_Instrumentation.java:207]:
```suggestion
```
I suspect it is a formatting rule for method signatures that are broken up onto multiple lines, as this one seems consistent. For the single line method signatures it does not.

**Comment by tbradellis** [comment]
Still need to get db instance info (host, port).

**Comment by tbradellis** [comment]
moved to new pr branch here:
https://github.com/newrelic/newrelic-java-agent/pull/1164

**Pull request closed without merging**